R/tweak-initial-estimates.R

```diff
@@ -0,0 +1,192 @@
+#'   inherit_param_estimates() %>% tweak_initial_estimates(.p = 0.2)
+#'
+#' # If you want to set the seed for reproducible results:
+#' mod2 <- withr::with_seed(1234, {
+#'    tweak_initial_estimates(mod2, .p = 0.2, digits = 3)
+#' })
+#'
+#' }
+#' @export
+tweak_initial_estimates <- function(
+    .mod,
+    .p = 0.1,
+    tweak = c("theta"),
+    digits = 3
+){
+
+  # Assertions
+  test_nmrec_version(.min_version = "0.3.0.8001")
+  check_model_object(.mod, "bbi_nonmem_model")
+  checkmate::assert_true(all(tweak %in% BBR_ESTIMATES_INHERIT))
+
+  # Get initial estimates
+  initial_est <- get_initial_est(.mod, flag_fixed = TRUE)
+
+  # Rounding
+  fmt_digits <- paste0("%.",digits,"G")
+
+  # nmrec model objects
+  mod_path <- get_model_path(.mod)
+  mod_lines <- nmrec::read_ctl(mod_path)
+
+  # Tweak each record type
+  # Note: the `is_empty` check is mainly for models that dont specify SIGMA
+  # records. This allows users to not need to change the `tweak` argument if the
+  # record type is missing from the control stream file.
+  # It is also useful for testing.
+
+  if("theta" %in% tweak && !rlang::is_empty(initial_est$thetas)){
+    # Tweak THETA
+    new_thetas <- tweak_thetas(initial_est$thetas, .p, digits)
+    # Update THETA Block
+    nmrec::set_theta(
+      mod_lines, values = new_thetas, bounds = "keep",
+      fmt = fmt_digits
+    )
+  }
+
+
+  # Write out mod_lines to model
+  nmrec::write_ctl(mod_lines, mod_path)
+
+  return(.mod)
+}
+
+#' Tweak values
+#'
+#' @param values vector of values to tweak
+#' @inheritParams tweak_initial_estimates
+tweak_values <- function(values, .p){
+
+  # Sample percentages & Preserve seed if set
+  tweak_perc <- stats::runif(length(values), -.p, .p)
+
+  # Tweak values
+  new_values <- values + (values * tweak_perc)
+
+  # Return tweaked values
+  return(new_values)
+}
+
+
+
+
+#' Tweak theta estimates while respecting bounds
+#'
+#' @param init_thetas table of theta estimates and bounds
+#' @inheritParams tweak_initial_estimates
+#'
+#' @keywords internal
+tweak_thetas <- function(init_thetas, .p, digits){
+
+  init_thetas$new <- withr::with_preserve_seed(tweak_values(init_thetas$init, .p))
+
+  # Ignore fixed values
+  init_thetas <- init_thetas %>% dplyr::mutate(
+    new = ifelse(fixed == TRUE, .data$init, .data$new)
+  )
+
+  fmt_digits <- paste0("%.",digits,"G")
+
+  ### Respect Bounds ###
+  # Calculate range of the bounds
+  init_thetas$upper_diff <- init_thetas$up - init_thetas$init
+  init_thetas$lower_diff <- init_thetas$init - init_thetas$low
+
+  # Initialize formatted version of initial value and bounds
+  init_thetas_adj <- init_thetas %>% mutate(
+    new_fmt = sprintf(fmt_digits, .data$new),
+    up_fmt = sprintf(fmt_digits, .data$up),
+    low_fmt = sprintf(fmt_digits, .data$low)
+  )
+
+  # Set to `bound_perc` (e.g, 90%) of the way between init and bound
+  adjust_theta_bounds <- function(init_thetas, bound_perc = 0.9, fmt_digits){
```

Comment:
Small thing, but I don't think we want a default value for `bound_perc`. We'll always be passing something to this and this default just feels like it's asking for mysterious unintended behavior.